File: apps/api/app/schemas/billing.py
"""
Pydantic schemas for billing API endpoints.

These schemas define the request/response models for billing operations
including invoices, usage records, payments, and company billing settings.
"""

from datetime import datetime, date
from decimal import Decimal
from typing import List, Optional, Dict, Any
from uuid import UUID
from pydantic import BaseModel, Field, validator, condecimal

from app.models.invoice import InvoiceStatus
from app.models.company import PlanType
from app.core.pricing import PricingConstants

Money = condecimal(max_digits=18, decimal_places=2)


# Base schemas
class TimestampMixin(BaseModel):
    created_at: datetime
    updated_at: Optional[datetime] = None


# Company billing schemas
class CompanyBillingInfo(BaseModel):
    """Company billing information."""
    id: UUID
    name: str
    plan: PlanType
    quota_limit: Optional[int] = None
    quota_used: int = 0
    quota_remaining: Optional[int] = None
    billing_email: Optional[str] = None
    payment_customer_id: Optional[str] = None

    class Config:
        orm_mode = True


class CompanyBillingUpdate(BaseModel):
    """Update company billing settings."""
    plan: Optional[PlanType] = None
    quota_limit: Optional[int] = None
    billing_email: Optional[str] = None


class UsageStats(BaseModel):
    """Company usage statistics."""
    company_id: UUID
    current_month: int = 0
    current_week: int = 0
    today: int = 0
    total_usage: int = 0
    total_cost: Money = Field(default=Decimal("0"))
    quota_limit: Optional[int] = None
    quota_used: int = 0
    quota_remaining: Optional[int] = None


# Usage record schemas
class UsageRecordBase(BaseModel):
    """Base usage record schema."""
    action: str
    cost: Money
    session_id: Optional[UUID] = None
    metadata: Optional[Dict[str, Any]] = None


class UsageRecordCreate(UsageRecordBase):
    """Create usage record request."""
    pass


class UsageRecord(UsageRecordBase, TimestampMixin):
    """Usage record response."""
    id: UUID
    company_id: UUID
    user_id: Optional[UUID] = None

    class Config:
        orm_mode = True


class UsageRecordList(BaseModel):
    """Paginated list of usage records."""
    records: List[UsageRecord]
    total: int
    page: int
    per_page: int
    pages: int


# Invoice schemas
class InvoiceLineItemBase(BaseModel):
    """Base invoice line item schema."""
    description: str
    quantity: int = 1
    unit_price: Money
    amount: Money


class InvoiceLineItem(InvoiceLineItemBase):
    """Invoice line item response."""
    id: UUID

    class Config:
        orm_mode = True


class InvoiceBase(BaseModel):
    """Base invoice schema."""
    amount: Money
    currency: str = "BDT"
    due_date: date
    description: Optional[str] = None


class InvoiceCreate(InvoiceBase):
    """Create invoice request."""
    line_items: List[InvoiceLineItemBase] = []

    @validator('line_items')
    def validate_line_items(cls, v):
        if not v:
            raise ValueError("Invoice must have at least one line item")
        return v


class Invoice(InvoiceBase, TimestampMixin):
    """Invoice response."""
    id: UUID
    company_id: UUID
    invoice_number: str
    status: InvoiceStatus
    issued_date: date
    paid_date: Optional[date] = None
    payment_intent_id: Optional[str] = None
    payment_method: Optional[str] = None
    line_items: List[InvoiceLineItem] = []

    class Config:
        orm_mode = True


class InvoiceList(BaseModel):
    """Paginated list of invoices."""
    invoices: List[Invoice]
    total: int
    page: int
    per_page: int
    pages: int


class InvoiceUpdate(BaseModel):
    """Update invoice request."""
    status: Optional[InvoiceStatus] = None
    due_date: Optional[date] = None
    description: Optional[str] = None


# Payment schemas
class PaymentIntentCreate(BaseModel):
    """Create payment intent request."""
    invoice_id: Optional[UUID] = None
    amount: Optional[Money] = None
    currency: str = "BDT"
    payment_method_types: Optional[List[str]] = None
    return_url: Optional[str] = None
    cancel_url: Optional[str] = None

    @validator('currency')
    def validate_currency(cls, v):
        supported = ["BDT", "USD"]
        if v.upper() not in supported:
            raise ValueError(f"Currency must be one of: {', '.join(supported)}")
        return v.upper()


class PaymentIntent(BaseModel):
    """Payment intent response."""
    id: str
    amount: Money
    currency: str
    status: str
    client_secret: Optional[str] = None
    checkout_url: Optional[str] = None
    payment_method_types: List[str]
    created_at: datetime
    expires_at: Optional[datetime] = None


class PaymentResult(BaseModel):
    """Payment result response."""
    success: bool
    payment_id: str
    transaction_id: Optional[str] = None
    status: str
    amount: Money
    currency: str
    payment_method: Optional[str] = None
    error_message: Optional[str] = None


class RefundCreate(BaseModel):
    """Create refund request."""
    amount: Optional[Money] = None
    reason: Optional[str] = None


class RefundResult(BaseModel):
    """Refund result response."""
    success: bool
    refund_id: str
    original_payment_id: str
    amount: Money
    status: str
    error_message: Optional[str] = None


# Webhook schemas
class WebhookEvent(BaseModel):
    """Webhook event response."""
    event_id: str
    event_type: str
    payment_id: Optional[str] = None
    transaction_id: Optional[str] = None
    status: Optional[str] = None
    amount: Optional[Money] = None
    currency: Optional[str] = None
    timestamp: datetime
    is_verified: bool = False


# Pricing and quota schemas
class PricingInfo(BaseModel):
    """Current pricing information."""
    per_check: Decimal = PricingConstants.PER_CHECK
    import_draft: Decimal = PricingConstants.IMPORT_DRAFT
    import_bundle: Decimal = PricingConstants.IMPORT_BUNDLE
    currency: str = "BDT"

    class Config:
        use_enum_values = True


class QuotaCheck(BaseModel):
    """Quota check request."""
    action: str
    quantity: int = 1


class QuotaCheckResult(BaseModel):
    """Quota check result."""
    allowed: bool
    remaining: Optional[int] = None
    limit: Optional[int] = None
    message: Optional[str] = None


# Admin schemas (for admin endpoints)
class AdminCompanyStats(BaseModel):
    """Admin view of company statistics."""
    company_id: UUID
    company_name: str
    plan: PlanType
    total_usage: int
    total_cost: Money
    quota_limit: Optional[int] = None
    quota_used: int = 0
    last_activity: Optional[datetime] = None
    status: str  # active, suspended, etc.


class AdminUsageReport(BaseModel):
    """Admin usage report."""
    period_start: date
    period_end: date
    total_companies: int
    total_usage: int
    total_revenue: Money
    companies: List[AdminCompanyStats]


# Error schemas
class ErrorDetail(BaseModel):
    """Error detail schema."""
    message: str
    error_code: Optional[str] = None
    details: Optional[Dict[str, Any]] = None


class ValidationError(BaseModel):
    """Validation error schema."""
    field: str
    message: str
    value: Any


class HTTPError(BaseModel):
    """HTTP error response schema."""
    detail: str
    errors: Optional[List[ValidationError]] = None

File: apps/web/src/types/billing.ts
/**
 * TypeScript types for billing system
 */

export enum PlanType {
  FREE = 'FREE',
  STARTER = 'STARTER',
  PROFESSIONAL = 'PROFESSIONAL',
  ENTERPRISE = 'ENTERPRISE'
}

export enum InvoiceStatus {
  PENDING = 'PENDING',
  PAID = 'PAID',
  OVERDUE = 'OVERDUE',
  CANCELLED = 'CANCELLED',
  FAILED = 'FAILED'
}

export enum SettlementStatus {
  PENDING = 'PENDING',
  PROCESSED = 'PROCESSED',
  RECONCILED = 'RECONCILED',
  DISPUTED = 'DISPUTED'
}

export enum PaymentProvider {
  SSLCOMMERZ = 'sslcommerz',
  STRIPE = 'stripe'
}

export enum Currency {
  BDT = 'BDT',
  USD = 'USD'
}

export interface CompanyBillingInfo {
  id: string;
  name: string;
  plan: PlanType;
  quota_limit: number | null;
  quota_used: number;
  quota_remaining: number | null;
  billing_email: string | null;
  payment_customer_id: string | null;
}

export interface UsageStats {
  company_id: string;
  current_month: number;
  current_week: number;
  today: number;
  total_usage: number;
  total_cost: number;
  quota_limit: number | null;
  quota_used: number;
  quota_remaining: number | null;
}

export interface UsageRecord {
  id: string;
  company_id: string;
  user_id: string | null;
  session_id: string | null;
  action: string;
  cost: number;
  created_at: string;
  updated_at: string | null;
  metadata: Record<string, any> | null;
}

export interface UsageRecordList {
  records: UsageRecord[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

export interface InvoiceLineItem {
  id: string;
  description: string;
  quantity: number;
  unit_price: number;
  amount: number;
}

export interface Invoice {
  id: string;
  company_id: string;
  invoice_number: string;
  amount: number;
  currency: Currency;
  status: InvoiceStatus;
  issued_date: string;
  due_date: string;
  paid_date: string | null;
  payment_intent_id: string | null;
  payment_method: string | null;
  description: string | null;
  line_items: InvoiceLineItem[];
  created_at: string;
  updated_at: string | null;
}

export interface InvoiceList {
  invoices: Invoice[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

export interface PaymentIntent {
  id: string;
  amount: number;
  currency: Currency;
  status: string;
  client_secret: string | null;
  checkout_url: string | null;
  payment_method_types: string[];
  created_at: string;
  expires_at: string | null;
}

export interface PricingInfo {
  per_check: number;
  import_draft: number;
  import_bundle: number;
  currency: Currency;
}

export interface QuotaCheckResult {
  allowed: boolean;
  remaining: number | null;
  limit: number | null;
  message: string | null;
}

// Plan definitions
export interface PlanFeatures {
  name: string;
  quota: number | null;
  price: number;
  currency: Currency;
  features: string[];
  popular?: boolean;
}

export const PLAN_DEFINITIONS: Record<PlanType, PlanFeatures> = {
  [PlanType.FREE]: {
    name: 'Free',
    quota: 5,
    price: 0,
    currency: Currency.BDT,
    features: ['5 LC validations per month', 'Basic validation features', 'Email support']
  },
  [PlanType.STARTER]: {
    name: 'Starter',
    quota: 100,
    price: 15000,
    currency: Currency.BDT,
    features: ['100 LC validations per month', 'Standard features', 'Priority email support', 'Usage analytics'],
    popular: true
  },
  [PlanType.PROFESSIONAL]: {
    name: 'Professional',
    quota: 500,
    price: 45000,
    currency: Currency.BDT,
    features: ['500 LC validations per month', 'Advanced features', 'Phone support', 'Custom integrations', 'Advanced analytics']
  },
  [PlanType.ENTERPRISE]: {
    name: 'Enterprise',
    quota: null, // unlimited
    price: 0, // custom pricing
    currency: Currency.BDT,
    features: ['Unlimited validations', 'All features included', 'Dedicated support', 'Custom workflows', 'SLA guarantees']
  }
};

// Admin types
export interface AdminCompanyStats {
  company_id: string;
  company_name: string;
  plan: PlanType;
  total_usage: number;
  total_cost: number;
  quota_limit: number | null;
  quota_used: number;
  last_activity: string | null;
  status: string;
}

export interface AdminUsageReport {
  period_start: string;
  period_end: string;
  total_companies: number;
  total_usage: number;
  total_revenue: number;
  companies: AdminCompanyStats[];
}

export interface SystemBillingKPIs {
  total_revenue_monthly: number;
  active_companies: number;
  delinquent_accounts: number;
  invoices_this_cycle: number;
  average_revenue_per_company: number;
  quota_utilization_average: number;
}

// API request types
export interface UsageRecordsFilters {
  page?: number;
  per_page?: number;
  start_date?: string;
  end_date?: string;
  action?: string;
  // Bank-specific filters
  client_id?: string;
  branch_id?: string;
  product?: string;
}

export interface InvoicesFilters {
  page?: number;
  per_page?: number;
  status?: InvoiceStatus;
  // Bank-specific filters
  client_id?: string;
  branch_id?: string;
  product?: string;
  settlement_status?: SettlementStatus;
}

export interface CheckoutRequest {
  plan: PlanType;
  provider: PaymentProvider;
  return_url?: string;
  cancel_url?: string;
  amount?: number;
  currency?: string;
  priceId?: string;
  quantity?: number;
  mode?: 'payment' | 'subscription';
  metadata?: Record<string, any>;
  paymentMethodTypes?: string[];
}

export interface CompanyBillingUpdate {
  plan?: PlanType;
  quota_limit?: number;
  billing_email?: string;
}

// Bank compliance types
export interface BankComplianceReport {
  period_start: string;
  period_end: string;
  total_companies: number;
  active_companies: number;
  total_usage: number;
  total_revenue: number;
  compliance_score: number;
  usage_trends: BankUsageTrend[];
  plan_distribution: PlanDistribution[];
  compliance_metrics: ComplianceMetric[];
}

export interface BankUsageTrend {
  month: string;
  total_usage: number;
  total_revenue: number;
  active_companies: number;
  compliance_score: number;
  avg_cost_per_company: number;
}

export interface PlanDistribution {
  plan: PlanType;
  companies: number;
  revenue: number;
  percentage: number;
}

export interface ComplianceMetric {
  metric: string;
  score: number;
  status: 'excellent' | 'good' | 'warning' | 'critical';
  trend: 'up' | 'down' | 'stable';
  description?: string;
}

export interface SMEMetrics {
  total_companies: number;
  active_companies: number;
  total_usage: number;
  total_revenue: number;
  compliance_score: number;
  avg_monthly_usage: number;
  top_plan: PlanType;
  usage_growth: number;
  revenue_growth: number;
  period: string;
}

export interface ComplianceFilters {
  period: '1m' | '3m' | '6m' | '12m';
  start_date?: string;
  end_date?: string;
  include_inactive?: boolean;
  plan_filter?: PlanType;
}

// Chart data types for analytics
export interface UsageTrendData {
  date: string;
  usage: number;
  cost: number;
}

export interface RevenueTrendData {
  date: string;
  revenue: number;
  plan_breakdown: Record<PlanType, number>;
}

export interface QuotaThresholds {
  warning: number; // percentage (e.g., 80)
  critical: number; // percentage (e.g., 95)
  exceeded: number; // percentage (100)
}

export const DEFAULT_QUOTA_THRESHOLDS: QuotaThresholds = {
  warning: 80,
  critical: 95,
  exceeded: 100
};

// Utility functions
export function formatCurrency(amount: number, currency: Currency = Currency.BDT): string {
  const symbols = {
    [Currency.BDT]: '৳',
    [Currency.USD]: '$'
  };

  return `${symbols[currency]}${amount.toLocaleString()}`;
}

export function getQuotaThreshold(used: number, limit: number | null): 'normal' | 'warning' | 'critical' | 'exceeded' {
  if (!limit) return 'normal';

  const percentage = (used / limit) * 100;

  if (percentage >= DEFAULT_QUOTA_THRESHOLDS.exceeded) return 'exceeded';
  if (percentage >= DEFAULT_QUOTA_THRESHOLDS.critical) return 'critical';
  if (percentage >= DEFAULT_QUOTA_THRESHOLDS.warning) return 'warning';
  return 'normal';
}

export function getPlanDisplayName(plan: PlanType): string {
  return PLAN_DEFINITIONS[plan].name;
}

export function isUnlimitedPlan(plan: PlanType): boolean {
  return PLAN_DEFINITIONS[plan].quota === null;
}

export function getInvoiceStatusColor(status: InvoiceStatus): string {
  switch (status) {
    case InvoiceStatus.PAID:
      return 'text-green-600 bg-green-100';
    case InvoiceStatus.PENDING:
      return 'text-yellow-600 bg-yellow-100';
    case InvoiceStatus.OVERDUE:
    case InvoiceStatus.FAILED:
      return 'text-red-600 bg-red-100';
    case InvoiceStatus.CANCELLED:
      return 'text-gray-600 bg-gray-100';
    default:
      return 'text-gray-600 bg-gray-100';
  }
}

// Bank-specific types
export interface BankContract {
  id: string;
  bank_id: string;
  contract_number: string;
  plan: PlanType;
  contract_term_months: number;
  start_date: string;
  end_date: string;
  quota_limit: number | null;
  overage_rate: number; // per validation
  billing_contact_name: string;
  billing_contact_email: string;
  billing_contact_phone: string | null;
  po_reference: string | null;
  next_settlement_date: string;
  payment_terms: string; // e.g., "Net-30", "Net-45"
  currency: Currency;
  status: 'active' | 'expired' | 'terminated';
  created_at: string;
  updated_at: string | null;
}

export interface Allocation {
  id: string;
  bank_id: string;
  client_id: string | null; // null for bank-wide allocation
  client_name: string | null;
  branch_id: string | null;
  branch_name: string | null;
  product: string | null; // e.g., "LC_VALIDATION", "RE_CHECK"
  budget_limit: number | null; // null for unlimited
  quota_limit: number | null;
  usage_current_period: number;
  usage_cost_current_period: number;
  remaining_budget: number | null;
  alerts_enabled: boolean;
  alert_threshold_percent: number; // e.g., 80 for 80% warning
  period_start: string;
  period_end: string;
  created_at: string;
  updated_at: string | null;
}

export interface AllocationList {
  allocations: Allocation[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

export interface AllocationUpdate {
  budget_limit?: number | null;
  quota_limit?: number | null;
  alerts_enabled?: boolean;
  alert_threshold_percent?: number;
}